### PR TITLE
Add Email Signup Modal to Article Pages

### PR DESCRIPTION
**Pull request by mattThousand**
### Notion Spec
- https://www.notion.so/Add-email-signup-modal-to-all-article-pages-502ce8ab96c34c1aab3ef143f09f6a7f?pvs=4

### Demo

https://github.com/Pirate-Wires/pirate-wires/assets/2637355/e28b2121-a7dd-4d7e-8558-d8f5a0dbf3c8

https://github.com/Pirate-Wires/pirate-wires/assets/2637355/77d5e6a3-84d2-4faf-a656-339596d605b6

### Testing Instructions
** Ensure that `SITE_URL` is set to `http://localhost:3000` in your `.env.local` (not `https://localhost:3000`)

1. Run the app in a desktop browser as a signed out user (use an incognito tab and clear localstorage)
2. Navigate to any article
3. Scroll down ~35% of the vertical length of the article
4. Verify that the newsletter sign up modal appears (floating up from the bottom), the background overlay fades in, and the button in the lower righthand corner showing the number of free articles remaining appears
5. Verify that as you scroll further, the modal is locked in the vertical center of the viewport
6. Click the `X` icon on the "Free articles remaining" button
7. Verify that the "Free articles remaining" button disappears 
8. Click `Continue` 
9. Verify that the modal and background overlay disappear
10. Reload the page or navigate to a different article 
11. Scroll to the bottom of the article 
12. Verify that the modal UI does not appear
13. Clear `localstorage`
14. Repeat steps 1-3
15. When the modal appears, enter your email and click "Sign In"
16. Complete the sign in flow
17. Verify that you are redirected to the article you were initially reading
18. Scroll to the bottom of the article 
19. Verify that the modal UI does not appear
20. Sign out
21. Navigate to any article
22. add the query parameter `_cio_id=123` to the url string and reload the page
23. Scroll to the bottom of the article 
24. Verify that the modal UI does not appear
25. Load a new incognito browser session
26. Implement the following diff:
```
diff --git a/app/(website)/p/[slug]/page.tsx b/app/(website)/p/[slug]/page.tsx
index 8af9efa..897b0f2 100644
--- a/app/(website)/p/[slug]/page.tsx
+++ b/app/(website)/p/[slug]/page.tsx
@@ -87,7 +87,7 @@ export default async function PostDefault({ params }) {

   if (!session && post.slug) {
     const header = headers();
-    const ipAddress = (header.get("x-forwarded-for") ?? "127.0.0.1").split(",")[0];
+    const ipAddress = "127.0.0.456";

     const viewedArticles = await getViewedArticles(ipAddress);
     freeViewedArticleCount = viewedArticles.length;
```
27. Navigate to any article
28. Scroll down far enough for the modal UI to appear
29. Verify that the text "2 free articles remaining" appears on the button in the lower right corner
30. Without dismissing the modal, navigate to a new article
31. Repeat step 28
32. Verify that the text "1 free article remaining" appears on the button in the lower right corner
33. Repeat 30
34. Verify that instead of the email signup modal, the `Continue Reading With a Free Trial` bottom bar appears
35. Load a new incognito browser session
36. Navigate to an individual article
37. Subscribe to the article
38. Navigate to another article from the same newsletter
39. Verify that the modal UI does not appear
40. Navigate to an article in a different newsletter
41. Verify that the modal UI appears

### CODE REVIEW

File: components/ui/Overlay/sticky/StickyOverlay.tsx [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+    event?.preventDefault();
+
+    setIsOpen(false);
+
+    localStorage.setItem("lastOverlayDismissal", new Date().getTime().toString());
+  };
+
+  const handleSuccess = () => {
+    const newsletterSubs = localStorage.getItem("newsletterSubs");
+    localStorage.setItem("newsletterSubs", (newsletterSubs ?? "").concat(publication));
```
Comment (mattThousand) [inline components/ui/Overlay/sticky/StickyOverlay.tsx:72]: a better way of determining if a user already subscribed to a given newsletter might be to follow the pattern we're using to determine how many free articles a user has left (stores this data in the backend, by ip address) 🤔

**Review by pdsouza — COMMENTED** [review]
@mattThousand I'm probably being dumb here since I'm new to this but I'm seeing this subscribe footer whenever I view an article.. how do I trigger the modal like in your videos?

Screenshot 2024-02-05 at 11 41 57 PM

UPDATE: Ok, I was just able to get it to work but I realized that I was falling into the last case `freeViewedArticleCount >= 3` even though I started from scratch with no local storage and an incognito tab.

File: components/ui/Overlay/sticky/StickyOverlay.tsx [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+  const [error, setError] = useState<ToastableError | null>(null);
+
+  const newsletterTitle = useMemo(() => unslugifyAndCapitalize(publication.replace("The ", "")), [publication]);
+
+  const handleClose = (event?: any) => {
+    event?.preventDefault();
+
+    setIsOpen(false);
+
+    localStorage.setItem("lastOverlayDismissal", new Date().getTime().toString());
```
Comment (pdsouza) [inline components/ui/Overlay/sticky/StickyOverlay.tsx:61]: nit: `Date.now().toString()` is a little clearer (and a little faster, not that it matters here lol)

File: components/ui/Overlay/sticky/StickyOverlay.tsx [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+      if (!updateResponse.ok) {
+        const data = await updateResponse.json();
+        setIsProgress(false);
+        setIsLoading(false);
+        throw new ToastableError(data.message, updateResponse.status);
+      }
+
+      const updateData = await updateResponse.json();
+      if (updateData.success) {
```
Comment (pdsouza) [inline components/ui/Overlay/sticky/StickyOverlay.tsx:106]: What if this is `false`? Should we handle the error somehow?

**Review by pdsouza — CHANGES_REQUESTED** [review]
Beast, if I start a new incognito tab with seemingly no local storage set I get the paywall as though `freeViewedArticleCount >= 3`

Screenshot 2024-02-06 at 12 03 36 AM

Keep me honest if I'm doing something dumb, I'm still a noob on this project

**Comment by mattThousand** [comment]
> Beast, if I start a new incognito tab with seemingly no local storage set I get the paywall as though `freeViewedArticleCount >= 3`
> 
> Screenshot 2024-02-06 at 12 03 36 AM
> Keep me honest if I'm doing something dumb, I'm still a noob on this project

@pdsouza `freeViewedArticleCount` is actually driven by ip address and stored on the server (https://github.com/Pirate-Wires/pirate-wires/blob/b619ec1f9a1b1d7c9d797da09052f81751657174/app/(website)/p/%5Bslug%5D/page.tsx#L88), so in order to create the state(s) needed for QAing, I think you either would need to modify that line or use a VPN 🧪 👨‍🏭

File: components/ui/Overlay/sticky/StickyOverlay.tsx (hunk 3 above)
Comment (mattThousand) [reply components/ui/Overlay/sticky/StickyOverlay.tsx:113]: good catch - added an error handing function 👍

**Review by pdsouza — APPROVED** [review]
Nice work man! Everything works according to instructions! 🥇🥇 🥇

File: components/ui/Overlay/sticky/StickyOverlay.tsx [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+        setIsSuccess(true);
+
+        event.target.email.value = "";
+      } else {
+        const e = new ToastableError("An error occurred");
+        handleError(e);
```
Comment (pdsouza) [inline components/ui/Overlay/sticky/StickyOverlay.tsx:119]: Ah one more thing - I think we need a `return` after here? I don't see the error toast unless I add it.

**Review by pdsouza — DISMISSED** [review]
Just one small thing around the error handler - rest is good to go!

**Review by sbilling — APPROVED** [review]

**Pull request merged**